### PR TITLE
feat: Augments the send-side bwe to record some stats.

### PR DESCRIPTION
**Pull request by gpolitis**
This PR introduces some bwe related stats in an effort to understand
how much the loss-based algorithm is impacting our bwe. We record how
much time and how many times we spend in loss-limited, loss-degraded and
loss-free states.

### CODE REVIEW

File: src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java [hunk 1]
```diff
@@ -66,6 +86,31 @@
...
+
+    /**
+     * send_side_bandwidth_estimation.cc
+     */
+    private static final int kDefaultBitrateThresholdKbps = 0;
+
+    /**
+     * Disable the loss experiment by default.
+     */
+    private static final float kDefaultLossExperimentPct = 0;
```
Comment (bgrozev) [inline src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java]: Pct for "percent"? If I'm reading right it's used as a probability (0-1, not 0-100)

File: src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java [hunk 2]
```diff
@@ -488,4 +566,144 @@ private synchronized void fireBandwidthEstimationChanged(
             second = b;
         }
     }
+
+    public class StatisticsImpl implements Statistics
```
Comment (bgrozev) [inline src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java]: Does this come from webrtc.org?

File: src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java [hunk 3]
```diff
@@ -488,4 +566,144 @@ private synchronized void fireBandwidthEstimationChanged(
...
+        @Override
+        public void update(long nowMs)
+        {
+            synchronized (SendSideBandwidthEstimation.this)
+            {
+                update(nowMs, null);
+            }
+        }
+
+        // This method is synchronized on SendSideBandwidthEstimation.this.
```
Comment (bgrozev) [inline src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java]: There's no synchronized block. Do you mean that it should always be called while holding the lock? If so, I would make the method synchronized. Or at least make it private.

File: src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java (hunk 2 above)
Comment (gpolitis) [reply src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java]: No

File: src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java (hunk 3 above)
Comment (gpolitis) [reply src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java]: > Do you mean that it should always be called while holding the lock?

Yes.

> If so, I would make the method synchronized.

👍

File: src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java (hunk 1 above)
Comment (gpolitis) [reply src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java]: Probability <> percent and in this particular case we care about the percentage, in some form. What this variable holds is the proportion of lost packets; I felt that kDefaultLossExperimentProportion would be too long and that percentage (abbreviated as Pct) is clear enough. I'll finally rename the variable to kDefaultLossExperimentProportion.

**Comment by gpolitis** [comment]
test this please

File: src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java (hunk 2 above)
Comment (bgrozev) [reply src/org/jitsi/impl/neomedia/rtp/sendsidebandwidthestimation/SendSideBandwidthEstimation.java]: In this case, can you add documentation, please?

**Pull request merged**